src/Tools/LeanCode.CodeAnalysis/Analyzers/EnsureCQRSHandlersFollowNamingConvention.cs

```diff
@@ -0,0 +1,184 @@
+    internal const string CommandHandlerSuffix = "CH";
+    internal const string QueryHandlerSuffix = "QH";
+    internal const string OperationHandlerSuffix = "OH";
+
+    private static readonly DiagnosticDescriptor CommandHandlerRule =
+        new(
+            DiagnosticsIds.CommandHandlersShouldFollowNamingConvention,
+            "Command handlers should follow naming convention",
+            MessageFormat,
+            Category,
+            DiagnosticSeverity.Warning,
+            isEnabledByDefault: true
+        );
+
+    private static readonly DiagnosticDescriptor QueryHandlerRule =
+        new(
+            DiagnosticsIds.QueryHandlersShouldFollowNamingConvention,
+            "Query handlers should follow naming convention",
+            MessageFormat,
+            Category,
+            DiagnosticSeverity.Warning,
+            isEnabledByDefault: true
+        );
+
+    private static readonly DiagnosticDescriptor OperationHandlerRule =
+        new(
+            DiagnosticsIds.OperationHandlersShouldFollowNamingConvention,
+            "Operation handlers should follow naming convention",
+            MessageFormat,
+            Category,
+            DiagnosticSeverity.Warning,
+            isEnabledByDefault: true
+        );
+
+    public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics { get; } =
+        ImmutableArray.Create(CommandHandlerRule, QueryHandlerRule, OperationHandlerRule);
+
+    public override void Initialize(AnalysisContext context)
+    {
+        context.EnableConcurrentExecution();
+        context.ConfigureGeneratedCodeAnalysis(
+            GeneratedCodeAnalysisFlags.Analyze | GeneratedCodeAnalysisFlags.ReportDiagnostics
+        );
+        context.RegisterSymbolAction(AnalyzeSymbol, SymbolKind.NamedType);
+    }
+
+    public static void AnalyzeSymbol(SymbolAnalysisContext context)
+    {
+        var type = (INamedTypeSymbol)context.Symbol;
+
+        var handlerType = GetCQRSHandlerType(type, out var implementationCount);
+
+        if (handlerType == null)
+        {
+            return;
+        }
+
+        var expectedName = GetExpectedCQRSHandlerName(type, handlerType.Value, implementationCount);
+
+        if (type.Name != expectedName)
+        {
+            var rule = GetRule(handlerType.Value);
+            var diagnostic = Diagnostic.Create(rule, type.Locations[0], type.Name, expectedName);
+            context.ReportDiagnostic(diagnostic);
+        }
+    }
+
+    internal static string GetExpectedCQRSHandlerName(
+        INamedTypeSymbol type,
+        HandlerType handlerType,
+        int implementationCount
+    )
+    {
+        var expectedSuffix = GetSuffix(handlerType);
+
+        if (implementationCount > 1)
+        {
+            return type.Name.EndsWith(expectedSuffix, StringComparison.InvariantCulture)
+                ? type.Name
+                : type.Name + expectedSuffix;
```

Comment:
What happens here? It is when there is one handler for a few CQRS objects? How would this blow up if one would have the same handler for a few different CQRS objects (like the same handler for a command and a query) - just a thought experiment.

Reply:
Yes it's for a case when handler implements for example multiple commands. If we mix up commands, queries, etc no diagnostic will be raised as `var handlerType = GetCQRSHandlerType(type, out var implementationCount);` will return null and analyzer will return without reporting anything.